### PR TITLE
MTP-1478 add capping check in amount form

### PR DESCRIPTION
**Pull request by Staberinde**

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+10.18.1
```
Comment (ushkarev) [inline .nvmrc:1]: can we make this just `10` to be consistent with other apps? we don't need to be super specific with the version

File: mtp_send_money/apps/send_money/forms.py [hunk 2]
```diff
@@ -200,8 +209,52 @@ class DebitCardAmountForm(SendMoneyForm):
...
+            #  logger.exception('Could not look up prisoner account balance')
+            #  raise ValidationError(self.error_messages['connection'], code='connection')
+        #  return self.cleaned_data
+
+    def is_account_balance_below_threshold(self):
+        prisoner_account_balance = self.lookup_prisoner_account_balance(
+            prisoner_number=self.data['prisoner_number']
+        )['combined_prisoner_account_balance']
+        logger.info('Prisoner account balance for %s is %s', self.data['prisoner_number'], prisoner_account_balance)
+        return prisoner_account_balance < self.PRISONER_CAPPING_THRESHOLD_IN_PENCE
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:243]: this needs to be `<=` otherwise someone can't even reach the cap

File: mtp_send_money/apps/send_money/forms.py [hunk 3]
```diff
@@ -200,8 +209,52 @@ class DebitCardAmountForm(SendMoneyForm):
...
+        session = self.get_api_session(reconnect=(tries != 0))
+        try:
+            return session.get('/prisoner_account_balance/', params=filters).json()
+        except TokenExpiredError:
+            pass
+        except RequestException as e:
+            if e.response.status_code != 401:
+                raise
+        if tries < self.max_lookup_tries:
+            self.lookup_prisoner_account_balance(tries=tries + 1, **filters)
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:255]: this line needs to `return`

File: mtp_send_money/apps/send_money/forms.py (hunk 2 above)
Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:243]: My interpretation of the acceptance criteria was that they shouldn't be able to reach £900, only £899.99. https://dsdmoj.atlassian.net/browse/MTP-1465 Lets discuss it with the team tomorrow

Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:249]: the cap is definitely £900, not £899.99. in fact the story also says £1000, but that's wrong

File: mtp_send_money/apps/send_money/forms.py [hunk 4]
```diff
@@ -200,8 +208,59 @@ class DebitCardAmountForm(SendMoneyForm):
             'max_decimal_places': _('Only use 2 decimal places'),
         }
     )
+    error_messages = {
+        'connection': _('This service is currently unavailable'),
+        'cap_exceeded': _("The Prisoner's Account has reached its limit. \n Please try again in a few days.")
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:213]: this won't actually wrap to 2 lines when output into html. if it's _absolutely_ necessary to wrap this line, you could do something like the custom error message on the prisoner details form

File: mtp_send_money/apps/send_money/forms.py (hunk 2 above)
Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:249]: also, the form variable is lower-cased so this line should throw an attribute error.
i think it's best to refer to the setting directly here rather than copying it onto the form class first – that way tests can override the setting using `override_settings`.

File: mtp_send_money/apps/send_money/forms.py [hunk 5]
```diff
@@ -200,8 +208,59 @@ class DebitCardAmountForm(SendMoneyForm):
...
+            prisoner_number=self.prisoner_number
+        )['combined_prisoner_account_balance']
+        logger.info('Prisoner account balance for %s is %s', self.prisoner_number, prisoner_account_balance)
+        prisoner_account_balance += self.data['amount']
+        return prisoner_account_balance < self.PRISONER_CAPPING_THRESHOLD_IN_PENCE
+
+    def lookup_prisoner_account_balance(self, tries=0, **filters):
+        session = self.get_api_session(reconnect=(tries != 0))
+        try:
+            return session.get('/prisoner_account_balance/', params=filters).json()
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:254]: [api-602](https://github.com/ministryofjustice/money-to-prisoners-api/pull/602) uses the path rather than GET params as it's more REST-y. so this would be `session.get(f'/prisoner_account_balance/{prisoner_number}/').json()`

File: mtp_send_money/apps/send_money/forms.py [hunk 6]
```diff
@@ -57,6 +63,8 @@ def serialise_to_session(self):
                 value = getattr(cls, 'serialise_%s' % field)(value)
             session[field] = value
 
+        for field in getattr(cls, 'additional_fields_to_deserialize', []):
+            session[field] = getattr(self, field)
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:67]: this doesn't look right. why is something being saved into the session?

File: mtp_send_money/apps/send_money/forms.py [hunk 7]
```diff
@@ -42,7 +43,12 @@ def get_value(f):
             }
         except KeyError:
             data = None
-        return cls(request=request, data=data)
+
+        extra_kwargs = {
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:47]: you don't need `additional_fields_to_deserialize` anymore – the specific form's init method already consumes the kwarg explicitly. the `SendMoneyForm` doesn't need any changes

File: mtp_send_money/apps/send_money/tests/test_views.py [hunk 8]
```diff
@@ -574,19 +581,51 @@ def test_empty_form_shows_errors(self):
...
+                rsps.GET,
+                api_url(
+                    '/prisoner_account_balance/?{querystring}'.format(
+                        querystring=urlencode([
+                            ('prisoner_number', self.prisoner_number),
+                        ])
+                    )
+                ),
+                json={
+                    'data': {
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/tests/test_views.py:610]: the response doesn't have a 'data' key, the whole response is `{'combined_account_balance': 80000}`

File: .nvmrc (hunk 1 above)
Comment (mattbee) [reply .nvmrc:1]: Definitiely, this was left on my machine uncommitted, and I hadn't even looked at it so I'll update that for sure

File: mtp_send_money/apps/send_money/forms.py (hunk 7 above)
Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:47]: It still needs to be passed in when instantiating the class on line 51, so it will need to be stored somewhere, and storing it on the session with everything else seems like the best place for it

File: mtp_send_money/apps/send_money/forms.py (hunk 6 above)
Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:67]: That's the way it's done elsewhere, see line 64

File: mtp_send_money/apps/send_money/forms.py (hunk 7 above)
Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:47]: you pass it in already using `get_form_kwargs` of the view each time having loaded it from `valid_form_data`

Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:47]: oh wait, you mean on subsequent pages don't you

File: mtp_send_money/apps/send_money/forms.py (hunk 6 above)
Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:67]: `prisoner_number` is the only thing in `additional_fields_to_deserialize` and that's already in the session, no?

File: mtp_send_money/apps/send_money/forms.py [hunk 9]
```diff
@@ -200,8 +209,64 @@ class DebitCardAmountForm(SendMoneyForm):
             'max_decimal_places': _('Only use 2 decimal places'),
         }
     )
+    error_messages = {
+        'connection': _('This service is currently unavailable'),
+        'missing_prisoner_number': _('This service is currently unavailable'),
+        'cap_exceeded': _("The Prisoner's Account has reached its limit. Please try again in a few days.")
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:215]: is this the latest error message? the prototype has “You can’t send money to this person’s account. It has reached its safe limit for now. Please follow up with them about it.”

File: mtp_send_money/apps/send_money/forms.py [hunk 10]
```diff
@@ -200,8 +209,64 @@ class DebitCardAmountForm(SendMoneyForm):
...
     serialise_amount = serialise_amount
     unserialise_amount = unserialise_amount
+    max_lookup_tries = 2
+    additional_fields_to_deserialize = ['prisoner_number']
+    shared_api_session_lock = threading.RLock()
+    shared_api_session = None
+
+    def __init__(self, *args, **kwargs):
+        self.prisoner_number = kwargs.pop('prisoner_number', None)
+        if not self.prisoner_number:
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:226]: in what situation will this happen? the previous form would be invalid so users would get redirected there

File: mtp_send_money/apps/send_money/utils.py [hunk 11]
```diff
@@ -21,7 +21,7 @@
 prisoner_number_re = re.compile(r'^[a-z]\d\d\d\d[a-z]{2}$', re.IGNORECASE)
 
 
-def get_api_session():
+def get_api_session(reconnect=None):
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/utils.py:24]: what does `reconnect` do here?

File: mtp_send_money/apps/send_money/forms.py [hunk 12]
```diff
@@ -200,8 +209,64 @@ class DebitCardAmountForm(SendMoneyForm):
...
+                raise ValidationError(self.error_messages['cap_exceeded'], code='cap_exceeded')
+        except (RequestException, OAuth2Error):
+            logger.exception('Could not look up prisoner account balance')
+            raise ValidationError(self.error_messages['connection'], code='connection')
+        return self.cleaned_data
+
+    def is_account_balance_below_threshold(self):
+        prisoner_account_balance_integer = self.lookup_prisoner_account_balance()['combined_account_balance']
+
+        assert isinstance(prisoner_account_balance_integer, int), \
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/forms.py:249]: not sure this assertion is necessary given that we control the api's response type. also the assertion is never caught so a user would see a "500 service unavailable" rather than the form's error message. but i suppose this is an edge case anyway so 🤷

File: mtp_send_money/apps/send_money/tests/test_views.py [hunk 13]
```diff
@@ -574,19 +580,115 @@ def test_empty_form_shows_errors(self):
...
+        SERVICE_CHARGE_PERCENTAGE=Decimal('0'),
+        SERVICE_CHARGE_FIXED=Decimal('0'),
+        PRISONER_CAPPING_THRESHOLD_IN_POUNDS=Decimal('900')
+    )
+    def test_if_prisoner_cap_is_breached_error_displayed(self):
+        self.choose_debit_card_payment_method()
+        self.fill_in_prisoner_details()
+
+        with self.patch_prisoner_details_check(), responses.RequestsMock() as rsps:
+            mock_auth(rsps)
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/tests/test_views.py:599]: is this needed? the other new tests either have it commented out or it's absent

File: mtp_send_money/apps/send_money/tests/test_views.py [hunk 14]
```diff
@@ -574,19 +580,115 @@ def test_empty_form_shows_errors(self):
...
+            response = self.client.post(self.url, data={'amount': '100'}, follow=True)
+            self.assertOnPage(response, 'check_details')
+            self.assertEqual(self.client.session.get('amount'), '100.00')
+
+    @override_settings(
+        SERVICE_CHARGE_PERCENTAGE=Decimal('20'),
+        SERVICE_CHARGE_FIXED=Decimal('50'),
+        PRISONER_CAPPING_THRESHOLD_IN_POUNDS=Decimal('900')
+    )
+    def test_prisoner_cap_is_calculated_without_including_service_charge(self):
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/tests/test_views.py:666]: good idea!

File: mtp_send_money/apps/send_money/forms.py (hunk 6 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:67]: You're right, it appears to be there already and the tests pass without it.  Do we have a test Gov Pay set up on the test site still running, since I'm getting service unavailable locally (clicking next from the amount page) - just so I can sanity check the test setup matches the implementation.

Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:67]: Or is this related to the next discussion?

Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:67]: had a chat with @Staberinde yesterday and he suggested keeping this so that `additional_fields_to_deserialize` can be used for saving additional non-form-field-things into the session. to keep a kind of symmetry

File: mtp_send_money/apps/send_money/tests/test_views.py (hunk 13 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/tests/test_views.py:599]: Yes, I've still not found out why, but having it exist in other tests causes this failure
```
AssertionError: Not all requests have been executed [('POST', 'http://localhost:8000/oauth2/token/')]
```

while removing it from here causes all 4 capping tests to fail.

Comment (ushkarev) [reply mtp_send_money/apps/send_money/tests/test_views.py:599]: that's a little strange – it implies to me that some state is not reset between different runs of these tests. i guess we need to keep it for now.
would you mind adding a comment pls?

File: mtp_send_money/apps/send_money/forms.py (hunk 9 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:215]: You're right - updated :)

File: mtp_send_money/apps/send_money/utils.py (hunk 11 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/utils.py:24]: We use it in forms.py#260 to clean up the retry of connecting to the API. However to use it we needed to set it as a kwarg in the definition of `get_api_session` - does that make sense?

Comment (ushkarev) [reply mtp_send_money/apps/send_money/utils.py:24]: but `reconnect` is not used in the body of this function so can't do anything. you'll notice in [forms.py#234](https://github.com/ministryofjustice/money-to-prisoners-send-money/pull/362/files#diff-66826115334e40f34b2f571c0af52afaR234) that `reconnect` is not actually passed into this function

Comment (ushkarev) [reply mtp_send_money/apps/send_money/utils.py:24]: the class method `get_api_session` however _does_ use `reconnect` argument. it's only this line where it's not needed

Comment (mattbee) [reply mtp_send_money/apps/send_money/utils.py:24]: Removing this gives us a `TypeError: get_api_session() got an unexpected keyword argument 'reconnect'` - so I reckon we must be getting the mocks a bit wrong in the tests, as in the test we use the utils `get_api_session` it looks like. So from my moderate understanding we should be able to remove this but will need to sort the test_forms out.

Comment (ushkarev) [reply mtp_send_money/apps/send_money/utils.py:24]: yup i get it. i haven't played around with it but what happens if you remove test_forms.py#598: `mocked_api_session.side_effect = get_api_session`?

Comment (mattbee) [reply mtp_send_money/apps/send_money/utils.py:24]: Cool, that sorts it if we also remove `mock_auth(rsps)` from line 599

File: mtp_send_money/apps/send_money/forms.py (hunk 7 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:47]: So this is OK then? Not 100% clear on this comment

Comment (ushkarev) [reply mtp_send_money/apps/send_money/forms.py:47]: yeah i think so

File: mtp_send_money/apps/send_money/forms.py (hunk 10 above)
Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:228]: @Staberinde Do you agree this conditional will always return False, so we can drop it?

Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:228]: Only if `prisoner_number` were not recorded in the session somehow. I'm happy enough for it to be removed

**Comment by ushkarev** [comment]
oh and i've remembered something from the jira story: @mattbee suggested using a second environment variable to have as a switch for this feature. otherwise we can't deploy this change until policy takes effect.

File: mtp_send_money/apps/send_money/forms.py [hunk 15]
```diff
@@ -200,8 +209,64 @@ class DebitCardAmountForm(SendMoneyForm):
             'max_decimal_places': _('Only use 2 decimal places'),
         }
     )
+    error_messages = {
+        'connection': _('This service is currently unavailable'),
+        'missing_prisoner_number': _('This service is currently unavailable'),
+        'cap_exceeded': _('You can’t send money to this person’s account. \
```
Comment (Staberinde) [inline mtp_send_money/apps/send_money/forms.py:215]: Any reason for using this formatting on the string?

Comment (mattbee) [reply mtp_send_money/apps/send_money/forms.py:215]: You mean breaking over multiple lines? If so just to appease the linter.

**Comment by mattbee** [comment]
> oh and i've remembered something from the jira story: @mattbee suggested using a second environment variable to have as a switch for this feature. otherwise we can't deploy this change until policy takes effect.

We realised the same on Friday! So we'll get the feature flag in today.

File: mtp_send_money/apps/send_money/forms.py (hunk 15 above)
Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:215]: I meant more that elsewhere we've added line breaks inside the brackets elsewhere but omitted them here

File: mtp_send_money/settings/base.py [hunk 16]
```diff
@@ -290,6 +290,11 @@
 
 CLOUD_PLATFORM_MIGRATION_MODE = os.environ.get('CLOUD_PLATFORM_MIGRATION_MODE', '')
 
+PRISONER_CAPPING_ENABLED = os.environ.get('PRISONER_CAPPING_ENABLED', False)
```
Comment (ushkarev) [inline mtp_send_money/settings/base.py:293]: environment variables are always strings so this setting won't ever be `True`, it will be`'True'`. i'd recommend doing something like `os.environ.get('PRISONER_CAPPING_ENABLED', '') == 'True'`

File: mtp_send_money/apps/send_money/tests/test_forms.py [hunk 17]
```diff
@@ -36,17 +37,31 @@ def test(self):
             setattr(cls, 'test_valid__%s' % data_set['name'], make_method(data_set['input_data']))
 
     @classmethod
+    @override_settings(
```
Comment (ushkarev) [inline mtp_send_money/apps/send_money/tests/test_forms.py:40]: i'm pretty this decorator probably needs to be on the `def test` a couple of lines down. otherwise the settings are only overridden while the methods are being made rather than when they're run.

Comment (mattbee) [reply mtp_send_money/apps/send_money/tests/test_forms.py:40]: Yeah, we're still working on this, Toby spotted this too.

File: mtp_send_money/apps/send_money/forms.py (hunk 15 above)
Comment (Staberinde) [reply mtp_send_money/apps/send_money/forms.py:215]: Basically what I'm proposing is: 
```
cap_exceeded': _(
            'You can’t send money to this person’s account. It has reached its safe limit for now. '
            'Please follow up with them about it.'
        )
```

File: mtp_send_money/settings/base.py (hunk 16 above)
Comment (Staberinde) [reply mtp_send_money/settings/base.py]: We've changed it to use the binary-like representation for booleans in environment variables that tends to be standard practise in the 'nix world. Let me know if you've got any concerns

Comment (ushkarev) [reply mtp_send_money/settings/base.py]: in this form, the flag is "truthy" if the environment variable is set to anything at all (including 'False' or '0') except an empty string. i'm ok with that, but do note that the type of `PRISONER_CAPPING_ENABLED` can either be a string or bool depending on the environment

Comment (ushkarev) [reply mtp_send_money/settings/base.py]: oops that comment i wrote was related to the snippet this thread is attached to, not what the code says _now_!

Comment (ushkarev) [reply mtp_send_money/settings/base.py]: i don't really have concerns using `PRISONER_CAPPING_ENABLED=0` and `PRISONER_CAPPING_ENABLED=1`. that'd works fine though it's unlike all other boolean settings in the deploy repo
